Android/app/src/main/java/app/intra/socks/OverrideSocksHandler.java

```diff
@@ -116,16 +116,18 @@ public void doConnect(Session session, CommandMessage commandMessage) throws IOE
     download.setBufferSize(BUFFER_SIZE);
 
     StatsListener listener = null;
+    Exception initialError = null;
     try {
       listener = runPipes(upload, download, remoteServerPort, SIMULATE_RESET);
+      initialError = listener.error;
       // We consider a termination event as a potentially recoverable failure if
       // (1) It is HTTPS
       // (2) Some data has been uploaded (i.e. the TLS ClientHello)
       // (3) No data has been received
       // (4) The socket was terminated by a TCP RST
       // In this situation, the listener's uploadBuffer should be nonempty.
       if (remoteServerPort == HTTPS_PORT && listener.uploadBytes > 0 && listener.downloadBytes == 0
-          && listener.error != null && RESET_MESSAGE.equals(listener.error.getMessage())
+          && initialError != null && RESET_MESSAGE.equals(initialError.getMessage())
```

Comment:
Can't we check the type of the error instead? Is there any guarantee the message is stable and the same across Android versions? Isn't the message subjected to i18n?

Reply:
Humm, it seems you get `java.net.SocketException`, which may not be fine grained enough.